routes/tools_routes.js

```diff
@@ -0,0 +1,133 @@
+const express = require('express');
+const quickselect = require('quickselect');
+
+const carddb = require('../serverjs/cards');
+const Filter = require('../dist/util/Filter');
+
+const CardRating = require('../models/cardrating');
+
+const router = express.Router();
+
+const MAX_RESULTS = 300;
+
+/* Gets k sorted minimum elements of arr. */
+/* Modifies arr. */
+function sortLimit(arr, k, keyF) {
+  keyF = keyF || (x => x);
+  const compareF = (x, y) => keyF(x) - keyF(y);
+  if (k < arr.length) {
+    quickselect(arr, k, 0, arr.length - 1, compareF);
+  }
+  const result = arr.slice(0, k);
+  result.sort(compareF);
+  return result;
+}
+
+function matchingCards(filter) {
+  const cards = carddb.allCards();
+  if (filter.length > 0) {
+    return cards.filter(card => Filter.filterCard({
+      details: card,
+    }, filter, /* inCube */ false));
+  } else {
+    return cards;
+  }
+}
+
+function makeFilter(filterText) {
+  if (!filterText || filterText.trim() === '') {
+    return {
+      err: false,
+      filter: [],
+    };
+  }
+
+  const tokens = [];
+  const valid = Filter.tokenizeInput(filterText, tokens) && Filter.verifyTokens(tokens);
+
+  return {
+    err: !valid,
+    filter: valid ? [Filter.parseTokens(tokens)] : [],
+  };
+}
+
+function topCards(filter, res) {
+  const cards = matchingCards(filter);
+  const nameMap = new Map();
+  for (const card of cards) {
+    if (nameMap.has(card.name)) {
+      nameMap.get(card.name).push(card);
+    } else {
+      nameMap.set(card.name, [card]);
+    }
+  }
+  const names = [...nameMap.keys()];
+  const versions = [...nameMap.values()].map(possible => {
+    // TODO: pull out and use notPromoOrDigitalId in cube_routes.js
+    let nonPromo = possible.find(card => !card.promo && !card.digital && card.border_color != 'gold');
+    return nonPromo || possible[0];
+  });
+
+  return CardRating.find({
+    'name': {
+      $in: names,
+    },
+  }).then(ratings => {
```

Comment:
How expensive is this DB call? Is it possible we should be using the npm library `mongoose-paginate`?

Reply:
The expensive part isn't the returning of results - it's the call itself, so pagination wouldn't help. I haven't found it to be a problem on local but creation of an index on the name field should make it perfectly fine performance-wise. If it's causing problems we can revisit.

Reply:
Actually, scratch that - pagination would do something, but we need to see all the results anyway so it's not useful here.